Count declared methods and fields

vRallev opened the pull request:
I broke down the PR into smaller commits. It might be easier to review them individually. This PR resolves #270. 

Overall this PR adds one new parameter `printDeclarations` to the extension. When turned on the report prints also declared methods in separate columns. This only works for AARs, because I couldn't find a good way to find the `.jar` file for APKs. 

Additionally, I've added support for the `java` and `java-library` plugin. There `printDeclarations` must be turned on and the report only contains the declarations and not references.

vRallev commented:
Interesting, I cannot reproduce the CI failure. Locally all the tests are green for me.
```
com.getkeepsafe.dexcount.JarFileSpec > test AAR method count FAILED
    java.lang.IndexOutOfBoundsException: fromIndex = -1
```

benjamin-bader commented:
Bravo, and thank you!  It'll take me some time to go over this; between work and family, I don't have a lot of time for OSS these days, so thanks in advance for your patience.

benjamin-bader commented:
I can't reproduce the failure locally either, but logically it must happen on line 281 (line numbers are weird here, probably due to inline functions).  The second call to `indexOfFirstAfter` takes the output of the first.  The function returns `-1` if no elements match the predicate, which appears to be what happened here.

I don't know if there's some difference with the format of `javap` output between travis and our local environments, but it's my hunch.

benjamin-bader commented on src/main/kotlin/com/getkeepsafe/dexcount/Tasks.kt:
```diff
@@ -165,20 +168,38 @@ abstract class DexMethodCountTaskBase: DefaultTask() {
-        val dataList = DexFile.extractDexData(file, config.dxTimeoutSec)
+        val dataList = if (isAndroidProject) DexFile.extractDexData(file, config.dxTimeoutSec) else emptyList()
+        val jarFile = when {
+            isAndroidProject && config.printDeclarations -> JarFile.extractJarFromAar(file)
```
This won't work for APKs, for which `isAndroidProject` will be true.  I see that `printDeclarations` is disallowed for applications, but IMO a more-robust check here would be `isAar` (for example, in case the policy in the plugin changes).

vRallev replied:
You're right, I make this more explicit.

benjamin-bader commented on src/main/kotlin/com/getkeepsafe/dexcount/PrintOptions.kt:
```diff
@@ -16,14 +16,16 @@
+    var printHeader: Boolean = false,
+    var orderByMethodCount: Boolean = false,
+    var maxTreeDepth: Int = Integer.MAX_VALUE,
+    var printDeclarations: Boolean = false,
```
Ah, answered my own question this morning, please ignore this comment.

benjamin-bader commented:
Thank you again for this contribution!  I have a few small changes that I'd like to see, and some larger potential changes to think about.  The nits are marked inline, but I'd encourage you to consider using a bytecode library, rather than relying on javap (particularly given the non-locally-reproducible test failure in CI).

vRallev commented on src/main/kotlin/com/getkeepsafe/dexcount/Plugin.kt:
```diff
@@ -209,6 +228,45 @@ abstract class TaskProvider(
+    }
+
+    protected fun checkPrintDeclarationsIsTrue() {
+        check(ext.printDeclarations) { "printDeclarations must be true for Java projects: $project" }
```
I'd like to be explicit and make people aware of it. But I'm also okay to change this. Up to you.

vRallev commented on src/main/kotlin/com/getkeepsafe/dexcount/Plugin.kt:
```diff
@@ -137,6 +140,14 @@ abstract class TaskProvider(
                 ext!!.libraryVariants
             }
 
+            project.plugins.hasPlugin(JavaPlugin::class.java) || project.plugins.hasPlugin(JavaLibraryPlugin::class.java) -> {
```
I wouldn't. If you use Kotlin on the JVM, then you must also apply the Java plugin AFAIK. I don't how this would change with Kotlin multiplatform, though. I'm a little hesitant.

benjamin-bader replied:
Yeah, that's what I was thinking.  I hadn't considered multiplatform, and I'm not inclined to accommodate it now either.

vRallev commented on src/main/kotlin/com/getkeepsafe/dexcount/PackageTree.kt:
```diff
@@ -42,54 +45,36 @@ class PackageTree(
     // A cached sum of this node and all children's method ref counts.
     // Set by `getMethodCount()`, and invalidated by adding new nodes.
-    private var methodTotal_: Int? = null
+    private var methodTotal_ = mutableMapOf<Type, Int>()
```
That is correct. The alternative would have been to duplicate all the fields, but I liked this approach better. I guess it's a personal taste.

benjamin-bader replied:
It's fine by me, I just wanted to make sure I understood what was happening :)

vRallev commented on src/main/kotlin/com/getkeepsafe/dexcount/PackageTree.kt:
```diff
@@ -277,12 +308,19 @@ class PackageTree(
-        if (opts.includeMethodCount) {
-            json.name("methods").value(methodCount)
+        if (opts.isAndroidProject) {
+            if (opts.includeMethodCount) {
```
Yes, right now java projects only print declared fields and methods and not referenced once. The reason is that you could apply this plugin in now to java only projects without the Android toolchain to compute method references. I still find declared method useful, though.

I agree that the visitor or any other pattern would be better than all the if checks. But I feel like this should be done in a separate PR to reduce the noise? Are you okay with that?

benjamin-bader replied:
Yes, I agree - that's a big-enough job on its own.

vRallev commented on src/main/kotlin/com/getkeepsafe/dexcount/Plugin.kt:
```diff
@@ -209,6 +228,45 @@ abstract class TaskProvider(
+            task.dependsOn(jarTask)
+            task.mustRunAfter(jarTask)
+
+            // But package should always imply that dexcount runs, unless configured not to.
```
Removed.

vRallev commented on src/main/kotlin/com/getkeepsafe/dexcount/SourceFile.kt:
```diff
@@ -200,6 +195,184 @@ internal class DexFile: Closeable {
     }
 }
 
+class JarFile(
```
Done

vRallev commented on src/main/kotlin/com/getkeepsafe/dexcount/SourceFile.kt:
```diff
@@ -32,46 +33,40 @@ import java.util.zip.ZipEntry
 import java.util.zip.ZipException
 import java.util.zip.ZipFile
 
+sealed class SourceFile : Closeable {
```
Done

vRallev commented on src/main/kotlin/com/getkeepsafe/dexcount/SourceFile.kt:
```diff
@@ -200,6 +195,184 @@ internal class DexFile: Closeable {
+             * }
+             */
+            val lines = Runtime.getRuntime()
+                .exec("javap -private $classFiles")
```
I would think so that javap includes synthetics, but haven't tested it myself. I would give a bytecode library a try.

vRallev commented:
Thanks for the review @benjamin-bader. I've addressed the nits and will take a look at bytecode library next week to replace javap. The good thing is that tests are now in place, so that should make it safer. Did I miss anything else that you'd like to see changed?

vRallev commented:
@benjamin-bader I've added a new commit that uses Javassist instead of relying on Javap. The code also got simpler. Please let me know what you think.

benjamin-bader commented on src/main/kotlin/com/getkeepsafe/dexcount/Tasks.kt:
```diff
@@ -215,12 +236,18 @@ abstract class DexMethodCountTaskBase: DefaultTask() {
-            out.println("Fields remaining in $filename:  $fieldsRemaining")
-            out.println("Classes remaining in $filename:  $classesRemaining")
+            if (isAndroidProject) {
+                out.println("Total methods hello in $filename: ${tree.methodCount} ($percentMethodsUsed% used)")
```
I think 'hello' is extraneous here?  Integration tests are failing due to it.

vRallev replied:
This probably was a misplaced cursor. I'm glad it's not a password :-)

benjamin-bader commented:
Thanks so much for this feature- there's a ton of good work here.  I'm looking forward to shipping it!

vRallev commented:
Thank you, appreciated a lot!

benjamin-bader commented:
FYI, I _finally_ pushed a non-snapshot release that contains this PR.  Thanks for your patience - this was a major version revision (our first breaking change), so it took me quite a bit longer to feel good about calling it "done".

Version 1.0.0 should be out to Maven Central any minute now... four and a half years later, we're finally there :)

vRallev commented:
No worries, congrats for the release!
